### ISSUE TITLE
floats don't round trip

### ISSUE THREAD

**Issue by jonenst** [factor/factor#1555]
Factor:

``` factor
18270379323.2339630126953125 dup number>string string>number = .
f
```

Python:

``` python
>>> eval(repr (18270379323.2339630126953125 )) == 18270379323.2339630126953125 
True
```

Factor:
These 2 floats avec the same string representation:

``` factor
18270379323.2339630126953125 [ . ] [ .h ] bi
18270379323.23396
1.103ff74ecef94p34

IN: scratchpad 18270379323.233959197998046875 [ . ] [ .h ] bi
18270379323.23396
1.103ff74ecef93p34
```

Python: different representations

``` python
>>> 18270379323.2339630126953125 
18270379323.233963
>>> float.hex (18270379323.2339630126953125 )
'0x1.103ff74ecef94p+34'
>>> 18270379323.233959197998046875
18270379323.23396
>>> float.hex (18270379323.233959197998046875)
'0x1.103ff74ecef93p+34'

```

We use "%.16g" for number>string (which is less smart than python's approach from some floats, see #1158)  but it's not even enough to round trip. "%.17g" would be enough, but then more floats would be ugly: 1.1 is "1.1" with "%16g" but would be "1.1000000000000001" with "%17g". For the two floats in this issue, they would become "18270379323.233963" and "18270379323.233959"

I'll have to reread http://bugs.python.org/issue1580 but using "%15g" means we get no surprises like 0.07 -> "0.07000000000000001" that are happening in the current state.

So basically we use "%16g" which doesn't roundrip and has surpries (0.07): both problems
We could use "%15g" for no surprise or "%17g" for roundtrip.

**Comment by jckarter** [comment]
Problem might also be in the bignum library Factor uses to parse float literals, or maybe I didn't do the rounding correctly.

**Comment by jonenst** [comment]
There may be additional problems there, but in this case that's not it:
Once you write 1.103ff74ecef94p34 and 1.103ff74ecef93p34 both as "18270379323.23396", you can't round trip. (Note: Python uses the following distinct strings:
18270379323.233963
18270379323.23396
)

"%16g" will have these problems because there are more floats than strings with 16 digits precision. "%17g" doesn't have these problems.

Also, to parse "18270379323.23396", we end up doing 1827037932323396 10000 /f which is a fixnum/f on 64bits platforms, so bignums are not involved. For "18270379323.233963", it's a bignum/f and the code outputs the correct rounded result.

**Comment by bjourne** [comment]
We don't use format strings in `primitive_format_float` so I assume with "%16g" you mean that the c++ output stream is set to 16 digits of precision?

I read through the Python bug thread and they solved their problem by adding this giant file: https://hg.python.org/cpython/file/default/Python/dtoa.c, replacing the C librarys builtin float formatting and parsing functions with their own.

**Comment by jonenst** [comment]
> We don't use format strings in primitive_format_float so I assume with "%16g" you mean that the c++ output stream is set to 16 digits of precision?

Yes, it was "%16g" (and so the c++ version was written to have the same behavior) until a week ago when #1363 was merged in.

> I read through the Python bug thread and they solved their problem by adding this giant file: https://hg.python.org/cpython/file/default/Python/dtoa.c, replacing the C librarys builtin float formatting and parsing functions with their own. 

Yes, they used David's Gay code, which already had the mode:
> mode:        0 ==> shortest string that yields d when read in and rounded to nearest.

It's just that you can use this only if you know that the "read in" part will correctly round to nearest, and IEEE754 doesn't guarantee that (only that you round correctly with 17 digits)

So I think we can:
  * use "%15g" for simplicity and pretty ouput (lack correctness)
  * use "%17g" for simplicity and ugly ouput (but with correctness)
  * or use the same approach as python (but this sounds very difficult..)
      * Python's adaptation of Gay's code ?
      * https://github.com/google/double-conversion/ this looks really promising ?

**Comment by jonenst** [comment]
Also, we already have our own cross-platform "read in" part which does correctly round to nearest, so we could also use:
  * "%15g" , read back in, check equals, and "%17g" if not equal

**Comment by jonenst** [comment]
the "%15g" then "%17g" might actually be the best tradeoff.. It would fix #1158 and this one. The performance hit is not so terrible:

```
gc [ parse-float-benchmark ] time
! Before: Running time: 2.757036083 seconds
! After: Running time: 4.207476382 seconds
```

On the plus side, we can now do 
```diff
diff --git a/extra/benchmark/parse-float/parse-float.factor b/extra/benchmark/parse-float/parse-float.factor
index 99f0511..9bfef00 100644
--- a/extra/benchmark/parse-float/parse-float.factor
+++ b/extra/benchmark/parse-float/parse-float.factor
@@ -6,7 +6,7 @@ CONSTANT: test-floats $[ 100,000 random-units ]
 
 : parse-float-benchmark ( -- )
     test-floats [
-        [ number>string string>number ] [ 1e-10 ~ t assert= ] bi
+        [ number>string string>number ] [ assert= ] bi
     ] each ;
 
 MAIN: parse-float-benchmark
```

```diff
diff --git a/core/math/parser/parser.factor b/core/math/parser/parser.factor
index 4664c8b..96a0cf8 100644
--- a/core/math/parser/parser.factor
+++ b/core/math/parser/parser.factor
@@ -558,9 +558,16 @@ M: ratio >base
         [ drop ] if
     ] 2bi ; inline
 
+: format-general ( x digits -- string )
+    [ "" -1 ] dip "" "C" format-float ; inline
+
+: format-short-float ( x -- string )
+    dup 15 format-general dup string>number pick =
+    [ nip ] [ drop 17 format-general ] if ; inline
+
 : float>base ( n radix -- str )
     {
-        { 10 [ "" -1 16 "" "C" format-float ] }
+        { 10 [ format-short-float ] }
         [ bin-float>base ]
     } case ; inline
```

**Comment by bjourne** [comment]
That idea, formatting twice with different precision and comparing, was proposed in the Python bug too. I think Tim Peters had a good argument on why it couldn't work. 

But if it works, then great! Float formatting speed isn't that important. And if it is, we can always try and optimize the code in various ways.

**Comment by jonenst** [comment]
@bjourne 
They first proposed to do "%16g" and comparing, but this produces ugly floats for 0.07. Then someone said that this was not the case for "%15g", but by that time, they had decided against this because they wanted it to be cross platform. But we don't have the cross platform problem because our reading word (string>number) is cross-platform (ie it doesn't use the system libc) and correctly rounds to nearest.

I think they also chose their method for performance reasons.

**Comment by jonenst** [comment]
And I guess also because "try with different precision and compare" feels like an ugly hack to software engineers. Since all this was so that some floats could be prettier, people starting suggesting "always do %17g".

**Comment by jonenst** [comment]
Unless I misunderstood something :)

**Comment by jonenst** [comment]
Hmm wait, it relies on correctly rounded output from the C++ on "%15g", which probably doesn't work on microsoft VisualStudio.. So windows users wouldn't round trip, but would still have pretty floats with this.

**Comment by jonenst** [comment]
http://www.exploringbinary.com/visual-c-plus-plus-strtod-still-broken/#comment-9403 gives hope

**Comment by erg** [comment]
If I accidentally run the VS2013 terminal on the Windows build machine then a lot of the float tests fail with representation failures. VS2015 really made a commitment to fix them all and I think it worked.

Another approach to exact floats would be to implement unums.
https://github.com/jrmuizel/pyunum

**Comment by jonenst** [comment]
So let's do the "%15g" then "%17g" comparison approach. Apparently it's already broken for "%17g" on windows so it's not breaking anything and it fixes 0.07 printed as 0.07000000000000001 . And if VS2015 lives up to it's reputation, it makes float round-trip like on other platforms.
Is that ok ?

**Comment by jonenst** [comment]
So, do we proceed with the "%15g" then "%17g" approach ? And make VS2015 a requirement?

**Comment by mrjbq7** [comment]
In the short term, that seems reasonable because it is more correct.

Longer term, I'd like to have float formatting implemented in Factor correctly, similar to I guess the way Python has done it.

**Comment by mrjbq7** [comment]
Regarding VS2015, that seems okay to me, but I'm not a Windows guy.  @erg?

**Comment by jonenst** [comment]
From the python thread:
> there's no way to do it that satisfies all of {cheap, easy, fast, correct}

and
> if you use only bigint operations very straightforwardly, correct rounding amounts to a dozen lines of obviously correct Python code.

So what kind of implementation do you see for factor ? I think the python solution is too complex to maintain.

**Comment by mrjbq7** [comment]
The "dozen lines" sounds pretty attractive, no?  All code needs maintenance so if it is really complex, well then maybe we keep the formatting solution above.

**Comment by mrjbq7** [comment]
When I said "implemented in Factor", I meant in Factor code, not C.

**Comment by jonenst** [comment]
Ok, I was just making sure of what you meant. I think writing a simple (ie not very optimized) in factor would be easy to maintain, but we would lose the speed (but it's still better than maintaining a 3000 line C file like python does) (and now, they don't want to switch to grisu https://bugs.python.org/issue12450 )

**Comment by bpollack** [comment]
@mrjbq7 VS2015 is a completely fine requirement—and likely desirable, since MS is *finally* breaking up the C runtime in a sane way for backcompat. Assuming we want to target the same operating systems that MS still supports (Win 7, 8, 8.1, and 10), targeting VS2015 is genuinely desirable.

**Comment by jonenst** [comment]
Regarding performance, this has some numbers : https://github.com/miloyip/dtoa-benchmark#results
On my machine, for 2 millions floats,
gay's (mode0, ie pretty) dtoa: 12seconds
factor (stringstream "%.16g"),: 38 seconds. (with about 2/3 of the time in the c++ (format-float) )

C
```C
 for (i=0; i<2048; i++) {
   for (j=0; j<9007199254740992L; j+= 9007199154740) {
     long dlong = (i << 53) | j;
     double d= *((double *) &dlong);
     char *fmt= dtoa(d, 0, 0, &decpt,&sign,&rve);
   }
 }
```
compiled with ```gcc -O2 -DLong=int -DIEEE_8087=1  dtoa.c test.c```

Factor
```factor
: writenums ( -- )
2048 iota
0 9007199254740992 9007199154740 <range> [
[ swap 53 shift bitor bits>double number>string drop ] with each
] curry each ;
```

Factor profile
```
depth   time ms  GC %  JIT %  FFI %   FT %
  12   44038.0   5.84   0.00  84.12   0.01 T{ thread f "Listener" ~curry~ ~quotation~ 37 ~box~ f t...
  13   39585.0   3.96   0.00  85.32   0.01   M\ float >base
  14   30284.0   0.68   0.00  99.27   0.01     (format-float)
  14    2491.0   0.00   0.00   0.00   0.00     index
  15    1207.0   0.00   0.00   0.00   0.00       M\ string nth-unsafe
  16     272.0   0.00   0.00   0.00   0.00         M\ fixnum integer>fixnum
  15     308.0   0.00   0.00   0.00   0.00       M\ string length
  15      74.0   0.00   0.00   0.00   0.00       integer?
  14    1413.0  23.85   0.00  91.15   0.00     <string>
  14    1270.0  39.76   0.00  85.12   0.00     double>bits
  14    1254.0  26.00   0.00  44.02   0.00     cut
  15    1099.0  29.66   0.00  50.23   0.00       subseq
  16     663.0  49.17   0.00  83.26   0.00         <string>
  15     147.0   0.00   0.00   0.00   0.00       head
  14     752.0  25.66   0.00  73.94   0.00     subseq
  15     607.0  31.80   0.00  91.60   0.00       <string>
  14     482.0   0.00   0.00  48.96   0.00     bignum=
  14     387.0   0.00   0.00   0.00   0.00     member?
  14     275.0   0.00   0.00   0.00   0.00     M\ string equal?
  15     212.0   0.00   0.00   0.00   0.00       M\ string hashcode*
  14     264.0   0.00   0.00   0.00   0.00     tail
  15     184.0   0.00   0.00   0.00   0.00       length
  15      74.0   0.00   0.00   0.00   0.00       M\ string length
  14     130.0   0.00   0.00   0.00   0.00     M\ float fp-nan?
  13    1438.0  27.47   0.00  95.55   0.00   fixnum-shift
  13    1242.0  20.21   0.00  40.98   0.00   append
  14     703.0  35.70   0.00  72.40   0.00     <string>
  13     751.0  21.17   0.00  78.43   0.00   bignum-bitor
  13     580.0  11.03   0.00  94.66   0.00   bits>double
  13     329.0  41.64   0.00  75.38   0.00   fixnum>bignum
```

**Comment by jonenst** [comment]
And double-conversion https://github.com/google/double-conversion/ takes 1.3 seconds
(note that in the two examples before, I made a mistake on the size of the mantissa, it's fixed here. Also don't loop on nans and denormals)

```C++
 for (i=1; i<2047; i++) {
   for (j=0; j<4503599627370496L; j+= 4007199154740) {
     long dlong = (i << 52) | j;
     double d= *((double *) &dlong);
     DoubleToStringConverter::DoubleToAscii(d, DoubleToStringConverter::SHORTEST, 0, buf, sizeof buf, &s, &length, &decpt);
   }
 }
```

**Comment by jonenst** [comment]
Someone has started work on this in 2017 for python : https://bugs.python.org/issue12450#msg288171

**Comment by jonenst** [comment]
> Hmm wait, it relies on correctly rounded output from the C++ on "%15g", which probably doesn't work on microsoft VisualStudio.. So windows users wouldn't round trip, but would still have pretty floats with this.
Actually it relies on correctly rounded output for *prettyness* but not for *correctness*. 
If some platform incorrectly rounds a %15g output, then our crossplatform correctly rounding input code will detect a mismatch and so we will use a %17g output (which IEE754 says has to be correct). So this means that on windows, some floats will be printed as %17g instead of what could have been a nicer shorter representation. But I suppose that the floats which are not correctly rounded are not pretty floats that have only 1 or 2 digits after the comma anyway so they are already ugly. We want to avoid printing 1.1 as 1.1000000000000001 but we don't care if
1.12345678912345 is printed as the slightly longer
1.1234567891234501
right ?

**Comment by mrjbq7** [comment]
Should be fixed now!
